### PR TITLE
Add support for KFAC-expand and KFAC-reduce for `Linear` modules

### LINKED ISSUE #50: Support KFAC-expand and KFAC-reduce for linear weight-sharing layers.
Specifically, for `nn.Linear` and `nn.Conv2d` (after #45) layers.

### PR DESCRIPTION
**Pull request by runame**
Resolves #50.

This PR is based on #62 which should be reviewed first. I will tackle #45 next and directly implement expand and reduce (will reuse most of the code for linear modules since it is almost identical).

### CODE REVIEW

**Comment by f-dangel** [comment]
Ping me after #62 is merged (I believe yo also need to set the target branch of this PR to `main`).

**Comment by runame** [comment]
@f-dangel I merged #62 and resolved merge conflicts with this branch. (I previously set the base branch to the branch from #62 so the diffs are better visible, it got automatically changed to `development` after merging #62.)

**Review by f-dangel — APPROVED** [review]
Looks great up to minor stuff, please feel free to merge after addressing.

File: curvlinops/kfac.py [hunk 1]
```diff
@@ -164,11 +183,24 @@ def __init__(
...
                 f"Must be 'batch' or 'batch+sequence' if loss_func.reduction != 'sum'."
             )
+        if loss_func.reduction == "sum" and loss_average is not None:
+            # reduction used in loss function will overwrite loss_average
+            warnings.warn(
+                f"Loss function uses reduction='sum', but loss_average={loss_average}."
+                " loss_average is set to None.",
+                stacklevel=2,
+            )
+            loss_average = None
```
[the remark this answers is not in the document]
Comment (runame) [reply curvlinops/kfac.py:193]: I also thought about this, but I think in practice it is extremely likely that the user simply wants to use the reduction specified in the loss function and just forgot to adjust the `loss_average` argument. Ideally, we would set the loss function reduction based on the `loss_average` argument (so there can't be a mismatch), but this would change the base API which is why I didn't implement it.

But I'm open to raise an exception here as well, what do you think?

[the remark this answers is not in the document]
Comment (f-dangel) [reply curvlinops/kfac.py:198]: I would prefer the more strict version (exception). We can always relax it later if we want.
